lib/service/connect.go

```diff
@@ -784,36 +789,74 @@ func (process *TeleportProcess) rotate(conn *Connector, localState auth.StateV2,
 
 // newClient attempts to connect directly to the Auth Server. If it fails, it
 // falls back to trying to connect to the Auth Server through the proxy.
-func (process *TeleportProcess) newClient(authServers []utils.NetAddr, identity *auth.Identity) (*auth.Client, error) {
+// proxyFallbackAddr optionally specifies the address of the proxy to use as a fallback
+// if reverse tunnel connection fails. This case helps make teleport work in situations
+// when the proxy/reverse tunnel is being multiplexed on a single port but HTTP proxy configuration
+// disallows SSH traffic.
+// See github.com/gravitational/teleport/issues/4141 for details.
+func (process *TeleportProcess) newClient(authServers []utils.NetAddr, identity *auth.Identity, opts ...authClientOption) (*auth.Client, error) {
 	directClient, err := process.newClientDirect(authServers, identity)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
+	var cltConfig authClientConfig
+	for _, opt := range opts {
+		opt(&cltConfig)
+	}
+
 	// Try and connect to the Auth Server. If the request fails, try and
 	// connect through a tunnel.
 	process.log.Debugf("Attempting to connect to Auth Server directly.")
 	_, err = directClient.GetLocalClusterName()
 	if err != nil {
+		process.log.WithError(err).Warn("Failed to connect to Auth Server directly.")
 		// Don't attempt to connect through a tunnel as a proxy or auth server.
 		if identity.ID.Role == teleport.RoleAuth || identity.ID.Role == teleport.RoleProxy {
 			return nil, trace.Wrap(err)
 		}
 
-		process.log.Debugf("Attempting to connect to Auth Server through tunnel.")
-		tunnelClient, err := process.newClientThroughTunnel(authServers, identity)
+		process.log.Debug("Attempting to connect to Auth Server through tunnel.")
+		// Discover address of SSH reverse tunnel server.
+		proxyAddr, err := process.findReverseTunnel(authServers)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
```

Comment:
I initially wanted to factor the environment variable away from this method so that I could impose this new logic whenever I needed and with the value I needed be it from environment or elsewhere. Also, I thought this behavioral change might not be needed/desired in other places where newClient is used.

I now [moved](https://github.com/gravitational/teleport/pull/6056/commits/3c95cc68fe140b6f76fe5796011c8d06145f37c0#diff-054e31280a78f4aea4759cd740f8d8bcecf563679d7fe6b79991fdb944e8c611R506-R513) the configuration to an attribute and set it to a value from environment by default on configuration application. This is easier to handle and will not flaky as temporarily setting an environment variable.

Reply:
That's reasonable.

Also, I think with this new approach, we don't need to do fallback anymore. Just check if the tunnel address is set in the environment and use it to override `proxyAddr`.

```go
tunnelAddr := process.Config.SSH.ProxyReverseTunnelFallbackAddr
if tunnelAddr != "" {
    proxyAddr := tunnelAddr.String()
}